IDP-455 Update README.md (#17)
workleap/wl-domain-event-propagation: README.md

## Getting started

> |Package| Download Link                                                                                        | Description                                                                |

Reviewer: Because we link all the packages in a table, I think there's no need to preserve the nuget badge) at the very beginning of the document (only keep the build badge)

|----|------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------|

> |Workleap.DomainEventPropagation.Abstractions| ![nuget](https://img.shields.io/nuget/v/Workleap.DomainEventPropagation.Abscratcions.svg?logo=nuget) | Contains abstractions that are used for publishing and recieveing messages |

Reviewer 2: ```suggestion
|Workleap.DomainEventPropagation.Abstractions| ![nuget](https://img.shields.io/nuget/v/Workleap.DomainEventPropagation.Abstractions.svg?logo=nuget) | Contains abstractions that are used for publishing and receiving events |
```

> |Workleap.DomainEventPropagation.Publishing| ![nuget](https://img.shields.io/nuget/v/Workleap.DomainEventPropagation.Publishing.svg?logo=nuget)   | Contains classes for message publishing                                    |

Reviewer 2: ```suggestion
|Workleap.DomainEventPropagation.Publishing| ![nuget](https://img.shields.io/nuget/v/Workleap.DomainEventPropagation.Publishing.svg?logo=nuget)   | Contains classes to publish events                                    |
```

> |Workleap.DomainEventPropagation.Subscription| ![nuget](https://img.shields.io/nuget/v/Workleap.DomainEventPropagation.Subscription.svg?logo=nuget) |  Contains classes for subscriptions                                        |

Reviewer 2: ```suggestion
|Workleap.DomainEventPropagation.Subscription| ![nuget](https://img.shields.io/nuget/v/Workleap.DomainEventPropagation.Subscription.svg?logo=nuget) |  Contains classes to subscribe to topics and receive events                                        |
```

### What does the Workleap.DomainEventPropagation package do?
* Provides list of existing Topics
* Serializes domain events

> * Manages publication fields

Reviewer: I'm not sure to understand what this means although I have some knowledge of the lib and Event Grid, could we add details and try to read this as someone that doesn't really know anything?

> * Provides webhook endpoint (.net core only)

Reviewer: ```suggestion
* Provides webhook endpoint (.NET 6+ only)
```

* Provides IDomainEvent and IDomainEventHandler<> interfaces
* Deserializes domain events
* Mediates call to DomainEventHandlers

> * Validates configurations

Reviewer: ```suggestion
* Validates configuration
```

* Handles topic subscriptions

### Using the Workleap.DomainEventPropagation.Publishing package to publish events

> When using dotnet core, you can register event propagation at startup in the service collection.

Reviewer: This is actually not true, we're targeting .NET Standard 2.0 so you can publish events from (almost) any version of .NET.

> ```

Reviewer: ```suggestion
```csharp
```

// Method 1: Lazily bind the options to a configuration section
services.AddEventPropagationPublisher();
services.AddOptions<EventPropagationPublisherOptions>().BindConfiguration(EventPropagationPublisherOptions.SectionName);

// Method 2: Set options values directly in C#
services.AddEventPropagationPublisher(opt =>
{

>   opt.TopicName = "<topic_name_to _publish_to>",

Reviewer: ```suggestion
  opt.TopicName = "<topic_name_to_publish_to>",
```

>   opt.TopicAccessKey = "<provided from keyVault>",

Reviewer: With the recent merge of [IDP-452](https://github.com/gsoft-inc/wl-domain-event-propagation/pull/19), you can also use RBAC and provide Azure token credentials instead of using an access key, I think it deserves a mention in this README.

  opt.TopicEndpoint = "<azure_topic_uri>"
});
```
Configuration is required. Configuration can be loaded from the appsettings file by passing the IConfiguration instance (see above). The topic access key should be stored securely in a key vault.
```json
// appsettings.json
{
  "EventPropagation": {
    "TopicName": "<topic_name_to_publish_to>",
    "TopicAccessKey": "<keyVault_provided_value>",
    "TopicEndpoint": "<azure_topic_uri>"
  }
}
```

> To publish an event, use the IEventPropagationClient interface (via injection). Use the PublishDomainEventAsync to publish the event. The required Subject field is a string description to provide context for the event.

Reviewer: ```suggestion
To publish an event, use the `IEventPropagationClient` interface (via dependency injection). Use the `PublishDomainEventAsync` to publish the event. The required subject parameter is a string description to provide context for the event.
```

> ```
> var domainEvent = new ExampleDomainEvent
> {
>     Id = Guid.NewGuid().ToString(),
>     Date = DateTime.UtcNow
> };
> 
> await this._eventPropagationClient.PublishDomainEventAsync(subject: "TestEventPublication", coolCandorEvent);
> ```

Reviewer: ```suggestion
```csharp
var domainEvent = new ExampleDomainEvent
{
    Id = Guid.NewGuid().ToString(),
    Date = DateTime.UtcNow
};

await this._eventPropagationClient.PublishDomainEventAsync(subject: "TestEventPublication", domainEvent);
```
```

#### Restrictions
* Your service can publish events on only 1 topic.
* The topic the service publishes to must be in the Topics class provided by the Workleap.EventPropagation package.

### Using the Workleap.EventPropagation package to subscribe to events

> When using dotnet core, you can register event propagation subscriptions at startup in the service collection.  To configure the subscriber, the list of subscribed topics is required.

Reviewer: ```suggestion
When using ASP.NET Core with .NET 6+, you can register event propagation subscriptions at startup in the service collection. To configure the subscriber, the list of subscribed topics is required.
```

> ```

Reviewer: ```suggestion
```csharp
```

services
    .AddEventPropagationSubscriber()
    .AddDomainEventHandlersFromAssembly(Assembly.GetExecutingAssembly())

>     .Configure(new[] { Topics.Organisation, Topics.Candor });

Reviewer: > .Configure(new[] { Topics.Organisation, Topics.Candor });

See the message I sent you

// or add a single domain event handler
.AddDomainEventHandler<ExampleDomainEventHandler>()
```

> An API endpoint must be made available to receive events from the EventGrid topic. The package provides this endpoint for dotnet core projects. It can be added like so.

Reviewer: ```suggestion
An API endpoint must be made available to receive events from the Event Grid topic. The package provides this endpoint for ASP.NET Core projects. It can be added like so:
```

```
services.AddControllers().PartManager.ApplicationParts.Add(new AssemblyPart(typeof(EventGridController).Assembly));
```

#### Define domain event handlers

You must define domain event handlers for events you wish to handle. You are not obligated to handle every possible event from a topic.

> Domain event handlers must implement the IDomainEventHandler<> interface.
> 
> ```

Reviewer: ```suggestion
Domain event handlers must implement the `IDomainEventHandler<>` interface.

```csharp
```

public class ExampleDomainEventHandler : IDomainEventHandler<ExampleDomainEvent>
{
    public Task HandleDomainEventAsync(ExampleDomainEvent domainEvent, CancellationToken cancellationToken)
    {
        return Task.CompletedTask;
    }
}
```
#### Restrictions

> * You may only define 1 domain event handler per domain event you wish to handle. If you would require more, use the 1 allowed domain event handler as a façade for multiple operations.

Reviewer: ```suggestion
* You may only define one domain event handler per domain event you wish to handle. If you would require more, use the single allowed domain event handler as a facade for multiple operations.
```

* DomainEventHandlers must have idempotent behavior (you could execute it multiple times for the same event and the result would always be the same).
* DomainEventHandlers should use the property DataVersion (when necessary)

### Exposing domain events

> To expose your service’s domain events, you must provide a NuGet package with all domain (class) definitions. Your project must reference the Workleap.EventPropagation package and your domain events must implement the IDomainEvent interface.

Reviewer 2: ```suggestion
To expose your service’s domain events, you must provide a NuGet package with all domain (class) definitions. Your project must reference the Workleap.DomainEventPropagation.Abstractions package and your domain events must implement the IDomainEvent interface.
```

> ```

Reviewer: ```suggestion
```csharp
```

public class ExampleDomainEvent : IDomainEvent
{
    public string CoolId { get; set; }

>     public DateTime CoolDate { get; set; }

Reviewer 2: Maybe remove the cool part of property names?

    public string DataVersion => "1";
}
```

It is recommended to use a seperate pipeline to publish domain event definitions. Any existing pipeline can be reused with minor tweaks to achieve this.

#### Restrictions
* Domain events must only contain Ids and types (any content is optional). In some rare cases, an immutable field could be added to the event, however this comes at risk of exposing PII information.
* Changes to domain events should be additive
* Domain events must implement the DataVersion property
* Domain event projects should follow the pattern Workleap.[Service].[DomainEvents] (Service would usually match the topic name).

> * EventGrid limits event size to 64 kb. This means a regular event could contain a maximum of about 1500 Guid ids. Keep in mind the size restriction when sending batch events. Do not hesitate to split a larger event into multiple, smaller domain events.

Reviewer: ```suggestion
* Event Grid limits event size to 64 kb. This means a regular event could contain a maximum of about 1500 Guid ids. Keep in mind the size restriction when sending batch events. Do not hesitate to split a larger event into multiple, smaller domain events.
```

> ## Getting started

Reviewer 2: We could probably have a first section line in https://github.com/gsoft-inc/gsoft-authentication-clientcredentialsgrant where we briefly outline the different packages and what they include and the lower in the readme we go into more details on each with examples.

### What does the Workleap.EventPropagation package do?
* Provides list of existing Topics
* Serializes domain events
* Manages publication fields
* Provides webhook endpoint (.net core only)
* Provides IDomainEvent and IDomainEventHandler<> interfaces
* Deserializes domain events
* Mediates call to DomainEventHandlers
* Validates configurations
* Handles topic subscriptions

> ### Using the Workleap.EventPropagation package to publish events

Reviewer 2: ```suggestion
### Using the Workleap.DomainEventPropagation.Publishing package to publish events
```

When using dotnet core, you can register event propagation at startup in the service collection.
```
services
    .AddEventPropagation()
    .Configure(this.Configuration);
```
Configuration is required. Configuration can be loaded from the appsettings file by passing the IConfiguration instance (see above). The topic access key should be stored securely in a key vault.
```
"EventPropagation": {
    "TopicName": "...",
    "TopicAccessKey": "...",
    "TopicEndpoint": "http://topicurl.domain.com/api/events"
  }
```

To publish an event, use the IEventPropagationClient interface (via injection). Use the PublishDomainEventAsync to publish the event. The required Subject field is a string description to provide context for the event.

```
var coolCandorEvent = new ExampleDomainEvent
{
    CoolId = Guid.NewGuid().ToString(),
    CoolDate = DateTime.UtcNow

> };

Reviewer 2: I'd personnaly remove the 'cool' parts in here and also remove mentions to candor since it's an OV thing

await this._eventPropagationClient.PublishDomainEventAsync(subject: "TestEventPublication", coolCandorEvent);
```

#### Restrictions
* Your service can publish events on only 1 topic.

> * The topic the service publishes to must be in the Topics class provided by the Workleap.EventPropagation package.

Reviewer 2: Workleap.EventPropagation is not an existing package since ours is called DomainEventPropagation and I'm not sure if the whole sentence is still relevant or not

Reviewer 2: This comment is still relevant, the whole sentence could be removed

#### Exceptions

In case of error while publishing an event, an exception of type EventPropagationPublishingException is thrown. It provides the TopicName, Subject and TopicHostName fields.

> If configuration is bad, an exception of type InvalidConfigurationException is thrown. If the provided topic is bad, an exception of type InvalidTopicException is thrown.

Reviewer 2: Don't think this is true anymore, haven't found instances of those exceptions in the code

> ### Using the Workleap.EventPropagation package to subscribe to events

Reviewer 2: ```suggestion
### Using the Workleap.DomainEventPropagation.Subscription package to subscribe to events
```

Reviewer 2: This is still relevant

When using dotnet core, you can register event propagation subscriptions at startup in the service collection.  To configure the subscriber, the list of subscribed topics is required.

```
services
    .AddEventPropagationSubscriber()
    .AddDomainEventHandlersFromAssembly(Assembly.GetExecutingAssembly())

>     .Configure(new[] { Topics.Organisation, Topics.Candor });

Reviewer 2: We could probably add examples for other signatures like AddDomainEventHandler

Reviewer 2: Don't think the configure extension method still exists either

```

An API endpoint must be made available to receive events from the EventGrid topic. The package provides this endpoint for dotnet core projects. It can be added like so.
```

> services.AddControllers().PartManager.ApplicationParts.Add(new AssemblyPart(typeof(EventGridController).Assembly));

Reviewer 2: Think this is the 'old' way of doing it, now we have the AddEventPropagationEndpoints extension method to register those endpoints

Reviewer 2: This comment is still relevant I think

Reviewer: Yep @vad3245 in order to write the right documentation you should try the library

```

#### Define domain event handlers

You must define domain event handlers for events you wish to handle. You are not obligated to handle every possible event from a topic.

Domain event handlers must implement the IDomainEventHandler<> interface.

```
public class ExampleDomainEventHandler : IDomainEventHandler<ExampleDomainEvent>
{
    public Task HandleDomainEventAsync(ExampleDomainEvent domainEvent)
    {
        return Task.CompletedTask;
    }
}
```
To register domain event handlers, you can either use the package’s extension to scan your assembly and register all found domainevent handlers, or register your favorite domain event handlers manually.
```
// dotnet core service collection extension: scan assembly for all domain event handlers
.AddDomainEventHandlersFromAssembly(Assembly.GetExecutingAssembly())

// dotnet core service collection extension: add single domain event handler
.AddDomainEventHandler<ExampleDomainEventHandler>()

> ```

Reviewer 2: This section seems a bit redundant with the one above, we could probably combine them?

#### Exceptions
* If configuration is invalid, an exception of type InvalidConfigurationException is thrown.

> * If provided topic names are unknown, an exception of type InvalidTopicException is thrown.

Reviewer 2: Again those don't seem to exist in the solution anymore
